src/lbaf/IO/configurationValidator.py

```diff
@@ -102,35 +146,49 @@ def validate(valid_schema: Schema, schema_to_validate: dict):
     def main(self):
         """ Main routine for the config validation. """
         # Validate skeleton
-        if self.is_valid(valid_schema=self.__skeleton, schema_to_validate=self.__config_to_validate):
+        if self.is_valid(
+            valid_schema=self.__skeleton,
```

Comment:
Does it give an error?

Reply:
Python interpreter can parse it, but this is an error: `PEP 8: E125 continuation line with same indent as next logical line`,
it looks like they are all arguments of the same function.
```Python3
        if self.is_valid(
            valid_schema=self.__skeleton,
            schema_to_validate=self.__config_to_validate):
            self.__logger.info("Skeleton schema is valid")
```
We have line length up to 120 chars. It could be easily visible even with resolution 1280x1024 or 1280x720.
Should be as it was:
```Python3
        if self.is_valid(valid_schema=self.__skeleton, schema_to_validate=self.__config_to_validate):
            self.__logger.info("Skeleton schema is valid")
```

Reply:
Yes but such a resolution is hard to attain on a laptop.
It is easier to read/understand the code when function parameters are aligned in consecutive lines than when crammed into a single line.

Reply:
I think standard is 1920px, I have 2560px
Below you can not tell what's inside if, becasue indentation is wrong
```Python3
        if self.is_valid(
            valid_schema=self.__skeleton,
            schema_to_validate=self.__config_to_validate):
            self.__logger.info("Skeleton schema is valid")
```

Reply:
But what do you mean by **wrong**? Is it **wrong** in the sense that it causes a parsing/syntax error?

Or is **wrong** in a more relaxed sense? Like "it does not look good"?

Reply:
This causes linter error and violates PEP8, which we agreed to follow.
`PEP 8: E125 continuation line with same indent as next logical line`

The other thing is looks really bad, because all 3 lines are idented identical, which is not right visually + it's an anti-pattern, please see below from two different sources

![Zrzut ekranu z 2022-05-12 08-05-56](https://user-images.githubusercontent.com/47597569/168003230-f8afd82f-dbc7-4c38-8715-fd384ed2db76.png)

![Zrzut ekranu z 2022-05-12 08-09-44](https://user-images.githubusercontent.com/47597569/168003402-3afb9d02-825f-4984-9ccd-e23a923a487e.png)

Reply:
Thanks for the reference -- so let me make sure I get this right: the anti-pattern here is the fact that the conditional block is aligned with the condition itself, right?
```
if clause_1 or \
    clause_2:
    conditional_function()
```
That is the anti-pattern, right?

If so the PEP8 states that it should read instead:
```
if clause_1 or \
    clause_2:
        conditional_function()
```
Is that so?

We need to be completely clear about what we mean by **wrong**: does this cause the code to **break** or does it only violate some PEP8 style rule. I find it problematic with Python (unlike with other languages) that there is **syntactically correct** code that is nonetheless considered **incorrect** -- this opens the door to all sorts of interpretations/misunderstandings/disagreements about how the code should be.

If we want to say PEP8 is a strict requirement, I am not against it, but it also opens the door to us having to followed all sorts of extraneous rules that oftentimes seem more arbitrary/subjective than anything less. 

Do you see my point?

Reply:
By the way I just noticed something else in the code here, which is a more fundamental problem:
```
        # Validate skeleton
        if self.is_valid(
            valid_schema=self.__skeleton,
            schema_to_validate=self.__config_to_validate):
                self.__logger.info("Skeleton schema is valid")
        else:
            self.__logger.error("Skeleton schema is invalid")
            self.validate(
                valid_schema=self.__skeleton,
                schema_to_validate=self.__config_to_validate)
```
In the `else` block, we have an error message that is not followed by a clean exit, but instead is followed by a subsequent call to the `validate()` method. This is against the practice throughout LBAF that is to `sys.exit(1)` whenever an error is encountered. 
So instead of calling `validate()` here the code should terminate in error. 

Or, do you want to rather consider this `else` clause as leading not to an **error**, but to a mere **warning**?

Reply:
> Thanks for the reference -- so let me make sure I get this right: the anti-pattern here is the fact that the conditional block is aligned with the condition itself, right?
> 
> ```
> if clause_1 or \
>     clause_2:
>     conditional_function()
> ```
> 
> That is the anti-pattern, right?
> 
> If so the PEP8 states that it should read instead:
> 
> ```
> if clause_1 or \
>     clause_2:
>         conditional_function()
> ```
> 
> Is that so?

Yes, but the best would be, so we use line ot it's full length (120 chars):
```Python3
if clause_1 or clause_2:
    conditional_function()
```


> We need to be completely clear about what we mean by wrong: does this cause the code to break or does it only violate some PEP8 style rule. I find it problematic with Python (unlike with other languages) that there is syntactically correct code that is nonetheless considered incorrect -- this opens the door to all sorts of interpretations/misunderstandings/disagreements about how the code should be.

> If we want to say PEP8 is a strict requirement, I am not against it, but it also opens the door to us having to obey to all sorts of extraneous rules that oftentimes seem more arbitrary/subjective than anything less.

> Do you see my point?

I am pushing on the `minimum`, so the code is consistent and IMHO it is better to use the width of the screen and better to have screen covered with code instead of scrolling all the time.

Reply:
> By the way I just noticed something else in the code here, which is a more fundamental problem:
> 
> ```
>         # Validate skeleton
>         if self.is_valid(
>             valid_schema=self.__skeleton,
>             schema_to_validate=self.__config_to_validate):
>                 self.__logger.info("Skeleton schema is valid")
>         else:
>             self.__logger.error("Skeleton schema is invalid")
>             self.validate(
>                 valid_schema=self.__skeleton,
>                 schema_to_validate=self.__config_to_validate)
> ```
> 
> In the `else` block, we have an error message that is not followed by a clean exit, but instead is followed by a subsequent call to the `validate()` method. This is against the practice throughout LBAF that is to `sys.exit(1)` whenever an error is encountered. So instead of calling `validate()` here the code should terminate in error.
> 
> Or, do you want to rather consider this `else` clause as leading not to an **error**, but to a mere **warning**? I am unsure as to what your intent is, but my point is that if we want to let the code proceed then it cannot be an **error**.
> 
> BTW the same question applies to all four such conditionals that are following each other in the `main()` function.

Please take a look at the unit tests. When interpreter goes to `else`, we know that `schema` is invalid, because we checked `f self.is_valid(..)`, then in else it validates wrong schema with `validate()`, this raise `SchemaError` and returns a message what is wrong and exits as well. This SchemaError is then chatched in the tests and output is checked if it's that what we expect. So all good here.

Reply:
Philippe, we already spoke about that, I am writing unit tests, I can not test things properly without raising an error or exception, catching it and checking the error message. I thought we agreed on that I will be changing that respectively when writing unit tests. Please mind that raising error/exception is equal to sys.exit(1), but it can be tested.

Reply:
I don't remember we discussed that, but I believe it if you say so ;)

The problem I am seeing with this approach right now is that we have to be sure that the code actually calling the routine where the exception is raised, will indeed sys.exit(1) was expected. So it introduces the risk that it's actually not done in the main calling code.

I just want to make sure that each time we log an error, the code will exit. And here we have the log in one function, and the exit in another one. That's what I find a bit fragile. But if you're confident this will be maintained correctly, so be it!

Reply:
Philippe, you can try it by yourself, just make sure Schema is not right, and you will see how it works.
raising Error/Exception will always exit, the only case is when you use `try` and `except`, but that's not the case here.

Reply:
My question was really the following:
- if an exception in raised in `function_1()` (in `file_1.py`)
- but `function_1()` is called by `function_2()` (in `file_2.py`), and `function_2()` is where the `sys.exit(1)` **if exception raised call to ** `function_1()`
==> then there is the risk that if someone modifies at a later time  `function_2()` (in `file_2.py`) and inadvertently removes the `sys.exit(1)` **if exception raised call to ** `function_1()`, then the desired behavior will be changed because an error condition will no longer occur.

Please pay attention to the intent I am trying to achieve here: what I am saying is that be dissociating the place where the exception is raised, from the place where the subsequent exit will occur, we are introducing a maintenance liability. 

My point is not to say that **currently* the behavior is not what we want. My point is that this separation between place of exception from place of exit is making the code harder to maintain.

Are you seeing what I am trying to explain/convey here? It is **not** about current behavior, it is about future risks because of separation of exception from exit.

Reply:
I have checked the code a little bit on last Friday and there are places where we logging error and then sys.exit, but this error actually never happends, because in that case the error will be first noticed by ConfigurationValidator, this is something which needs to be cleaned up.
Regarding your comment from above:
We need to cover as much as we can  with unit tests, and in these unit tests we need to try to raise the errors/exceptions and see what is really reachable and remove all unnecessary code. Once we have that, it's very easy, one doesn't have to think what is intended, because it will be all covered in unit tests.
Modus operandi:
- one creates branch, one checks unit tests (unit tests should pass)
- one changes code
- one pushes a commit
- one checks unit tests and from there one sees what's expected and what's checked by the tests

Assuming, I don't like to speak what happend if, because anything could have happend and this discussion could be endless. I would like to really check the code, write unit tests and this really would be our guide. Any failing test would show to one, that something is wrong and should be changed or fixed.

P.S.
Regarding `Schema` library and this `SchemaError`, current behaviour in LBAF is due to this library and how it is designed. We could raise and error, just by checking if `schema.is_valid()`, but this bring us nothing, but when we then `schema.validate()` it will tell us where is the error.